### PR TITLE
Expand on our Timestamp mock

### PR DESCRIPTION
**Pull request by AverageHelper**
# Description
Our mocked `Firebase.Timestamp` class now covers Firestore's API for `toDate`, `toMillis` getters, and `fromDate` and `fromMillis` constructors.

The corresponding Jest mock functions may be used to inject return values to these functions, like so:
```js
// ... import firebase, mockTimestampToMillis ...

mockTimestampToMillis.mockReturnValue(400);
console.log(new firebase.firestore.Timestamp().toMillis()); // prints '400'
```

## Related issues
Depends on #38 

## How to test
```
jest
```

### CODE REVIEW

**Comment by FrozenKiwi** [comment]
I'm not sure if this change is inlcuded in this changelist (or if it listed somewhere else) but there is an issue in the timestamp implementation:

```
  static fromDate(date) {
    return (
      mockTimestampFromDate(...arguments) ||
      new Timestamp(date.getSeconds(), date.getMilliseconds() * 1000000)
    );
  }
```

note `date.getSeconds` returns only the number of seconds since the last minute, not the full number of seconds since epoch.

A suggested change here: https://github.com/thecointech/firestore-jest-mock/commit/9bc4d4a8048e5b604dfeceb95d463d49cfd9cb10

(or I can add this as a PR once the subCollections PR has been merged)

**Comment by AverageHelper** [comment]
That's a good point @FrozenKiwi. I believe `date.getTime` returns the time since reference in milliseconds, right? I'll put up a commit with that instead.

**Review by sbatson5 — COMMENTED** [review]
Nice work! 🙌  Just a couple of comments

File: __tests__/transaction.test.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  test('it returns a Promise', () => {
+    const result = db.runTransaction(async () => {});
+
+    expect(result).toBeInstanceOf(Promise);
+    expect(mockRunTransaction).toHaveBeenCalled();
+  });
+
+  test('it returns quickly', async () => {
+    await db.runTransaction(async () => {});
+    expect(true).toBe(true);
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply __tests__/transaction.test.js:42]: I'm... not sure. I have another idea for something to test, so I'll exchange this for something useful.

File: __tests__/transaction.test.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    });
+    expect(mockSetTransaction).toHaveBeenCalled();
+  });
+
+  test('mockUpdate is accessible', async () => {
+    expect.assertions(4);
+    expect(mockUpdateTransaction).not.toHaveBeenCalled();
+    const ref = db.collection('some').doc('body');
+
+    await db.runTransaction(async transaction => {
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply __tests__/transaction.test.js:95]: You're right, it doesn't need to be `async`. The canonical `transaction.update` returns immediately, deferring the actual transaction operation to when the function actually returns.

File: mocks/firestore.js [hunk 3]
```diff
@@ -255,36 +268,62 @@ FakeFirestore.Timestamp = class {
...
+    );
+  }
+
+  static fromMillis(millis) {
+    const d = new Date(0);
+    d.setMilliseconds(millis);
+    return mockTimestampFromMillis(...arguments) || FakeFirestore.Timestamp.fromDate(d);
+  }
+
+  static now() {
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/firestore.js:310]: A good idea. I'll throw one in.

File: mocks/helpers/buildDocFromHash.js [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
   return {
-    exists: !!hash || false,
-    id: hash.id || 'abc123',
-    ref: hash._ref,
+    exists,
+    id: (hash && hash.id) || id || 'abc123',
+    ref: hash && hash._ref,
     data() {
+      if (!exists) {
+        // From Firestore docs: "Returns 'undefined' if the document doesn't exist."
```
[the remark this answers is not in the document]
Comment (AverageHelper) [reply mocks/helpers/buildDocFromHash.js:9]: I'll add it. The doc is here: https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentSnapshot#data

**Review by sbatson5 — APPROVED** [review]
🎉 Nice work 🙌

**Comment by AverageHelper** [comment]
Thanks! #40 is ready for review.

**Pull request merged**